happi/tests/test_cli.py

```diff
@@ -95,9 +124,214 @@ def test_search_z(happi_cfg):
     assert [r.device for r in res] == [r.device for r in res_cli]
 
 
-def test_search_z_range(happi_cfg):
+def test_search_z_range(happi_cfg, caplog):
     client = happi.client.Client.from_config(cfg=happi_cfg)
     res = client.search_range('z', 3.0, 6.0)
     res_cli = happi.cli.happi_cli(['--verbose', '--path', happi_cfg, 'search',
                                    'z=3.0,6.0'])
     assert [r.device for r in res] == [r.device for r in res_cli]
+    # test invalid range
+    with caplog.at_level(logging.ERROR):
+        res_cli = happi.cli.happi_cli(['--verbose', '--path', happi_cfg,
+                                       'search', 'z=6.0,3.0'])
+        assert "Invalid range, make sure start < stop" in caplog.text
+        caplog.clear()
+
+
+def test_both_range_and_regex_search(happi_cfg):
+    client = happi.client.Client.from_config(cfg=happi_cfg)
+    # we're only interested in getting this entry (TST_BASE_PIM2)
+    res = client.search_regex(z='6.0')
+    # we're going to search for z=3,7 name=TST_BASE_PIM2
+    # we should only get in return one entry, even though the z value found 2
+    res_cli = happi.cli.happi_cli(['--verbose', '--path', happi_cfg, 'search',
+                                   'name=TST_BASE_PIM2', 'z=3.0,7.0'])
+    assert [r.device for r in res] == [r.device for r in res_cli]
+
+
+@pytest.mark.parametrize("from_user, expected_output", [(
+    # Test add item - succeeding
+    ['HappiItem', 'happi_name', 'device_class', ['arg1', 'arg2'],
+        {'name': 'my_name'}, True, 'docs', 'y'],
+    [
+        "Please select a container, or press enter for generic "
+        "Ophyd Device container",
+        "Enter value for name, default=None, "
+        "enforce=re.compile('[a-z][a-z\\\\_0-9]{2,78}$')",
+        "Enter value for device_class, default=None, enforce=<class 'str'>",
```

Comment:
Yeah I see what you mean - we would easily break the tests when modifying something related to the enforce.... 
Should I change it for all of them that use `enforce` then?

Reply:
I think this is good enough that we can revisit it if/when this becomes an issue. It may be far down the line we decide to update those criteria, anyway.

Reply:
Ok sounds good then, it will be pretty easy changing it to using regular expressions anyway, after we modify the `expected` strings like you mentioned above, we would then do something similar to this to assert them:
```
reg = re.compile(expected)
assert reg.match(message)
```

Reply:
Yep - definitely!

Reply:
I don't think we should be matching log or print output at all in this test suite, due to the volatility it introduces. Unit testing CLI is hard for this reason. Regex would be better, but even better would be breaking up the cli modules and testing their parts independently. I think that's the direction we should go if we revisit the test suite later.

Reply:
Yes I agree that refactoring the cli module to be able to test each command & their parts independently  would be a much better approach. Should I take these tests out from here?

Reply:
I think the coverage is worthwhile and your efforts here were useful. Regardless, I don't mind deferring to @ZLLentz's decision here.

Reply:
I think we should merge this, and keep in mind the refactor if we revisit the test suite.